### PR TITLE
WCASHMERE-50

### PR DESCRIPTION
**Pull request by barrycann**

### CODE REVIEW

File: projects/cashmere/src/lib/input/phone-mask.directive.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  ngOnDestroy() {
+    this.sub.unsubscribe(); 
+  }
+
+  phoneValidate(){
+
+    this.sub = this._phoneControl.valueChanges.subscribe(data => {
+      
+      /**the most of code from @Günter Zöchbauer's answer.*/
```
Comment (tfoxley) [inline projects/cashmere/src/lib/input/phone-mask.directive.ts]: Do you want to keep this comment?

File: projects/cashmere/src/lib/input/phone-mask.directive.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      } else {
+        newVal = newVal.replace(/^(\d{0,3})(\d{0,3})(.*)/, '($1) $2-$3');
+      }
+     
+      this._phoneControl.setValue(newVal,{emitEvent: false});
+       //keep cursor the normal position after setting the input above.
+      this.renderer.selectRootElement('#tel').setSelectionRange(start,end);
+    //when typed value in input
+    } else{
+      // this.message = 'Typing...'; //Just console
```
Comment (tfoxley) [inline projects/cashmere/src/lib/input/phone-mask.directive.ts]: Please remove this commented out code.

File: projects/cashmere/src/lib/input/phone-mask.directive.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    } else if (newVal.length <= 6) {
+      newVal = newVal.replace(/^(\d{0,3})(\d{0,3})/, '($1) $2');
+    } else {
+      newVal = newVal.replace(/^(\d{0,3})(\d{0,3})(.*)/, '($1) $2-$3');
+    }
+    //check typing whether in middle or not
+    //in the following case, you are typing in the middle
+    if(preInputValue.length >= start){
+      //change cursor according to special chars.
+      // console.log(start+removedD);
```
Comment (tfoxley) [inline projects/cashmere/src/lib/input/phone-mask.directive.ts]: Please remove this commented out code.

File: projects/cashmere/src/lib/input/phone-mask.directive.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      if(removedD == '-'){
+        start = start +1;
+        end = end +1;
+      }
+      if(removedD == " "){
+          start = start +1;
+          end = end +1;
+        }
+      this._phoneControl.setValue(newVal,{emitEvent: false});
+      this.renderer.selectRootElement('#tel').setSelectionRange(start,end);
```
Comment (tfoxley) [inline projects/cashmere/src/lib/input/phone-mask.directive.ts]: This is likely just me overlooking something but is this relying on the user to have added an id of `tel` to their input that they added the directive to? If they don't add the id does it all fall apart?

File: projects/cashmere/src/lib/input/phone-mask.directive.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    } else{
+        this._phoneControl.setValue(newVal,{emitEvent: false});
+        this.renderer.selectRootElement('#tel').setSelectionRange(start+2,end+2); // +2 because of wanting standard typing
```
Comment (tfoxley) [inline projects/cashmere/src/lib/input/phone-mask.directive.ts]: Same as previous comment about `tel` id.

**Comment by barrycann** [comment]
I've made the requested changes and re-committed.

**Pull request merged**